planning.py

```diff
@@ -1346,6 +1344,160 @@ def result(problem, action):
         else:
             return problem
 
+    
+
+    def angelic_search(problem, hierarchy, initialPlan):
+        """
+	[Figure 11.8] A hierarchical planning algorithm that uses angelic semantics to identify and
+	commit to high-level plans that work while avoiding high-level plans that don’t. 
+	The predicate MAKING-PROGRESS checks to make sure that we aren’t stuck in an infinite regression
+	of refinements. 
+	At top level, call ANGELIC -SEARCH with [Act ] as the initialPlan .
+
+        initialPlan contains a sequence of HLA's with angelic semantics 
+
+        The possible effects of an angelic HLA in initialPlan are : 
+        ~ : effect remove
+        $+: effect possibly add
+        $-: effect possibly remove
+        $$: possibly add or remove
+	"""
+        frontier = deque(initialPlan)
+        n=0
+        while n<4: 
+            if not frontier:
+                return None
+            plan = frontier.popleft() # sequence of Nodes? (HLA/Angelic HLA as actions)
+            opt_reachable_set = Problem.reach_opt(problem.init, plan)
+            pes_reachable_set = Problem.reach_pes(problem.init, plan)
+            if problem.intersects_goal(opt_reachable_set): # all reachable states
+                if Problem.is_primitive( plan, hierarchy ): 
+                    return ([x for x in plan.action])
+                guaranteed = problem.intersects_goal(pes_reachable_set) 
+                if guaranteed and Problem.making_progress(plan, plan):
+                    final_state = guaranteed[0] # any element of guaranteed 
+                    #print('decompose')
+                    return Problem.decompose(hierarchy, problem, plan, final_state, pes_reachable_set)
+                hla = None # there should be at least one HLA/Angelic_HLA, otherwise plan would be primitive.
+                j=-1
+                for i in range(len(plan.action)):
+                    if type(plan.action[i]) is HLA or type(plan.action[i]) is Angelic_HLA:
+                        hla = plan.action[i] 
+                        j=i
+                        break
+
+                if j>0:
+                    prefix = plan.action[:j-1]
+                    outcome = Problem.result(problem, prefix[-1])
+                else: 
+                    prefix = []
+                    outcome = copy.deepcopy(problem)
+                if j < len(plan.action)-1:
+                    suffix = plan.action[j+1:]
+                else: 
+                    suffix = []
```

Comment:
Please put the code that computes the prefix and suffix in a separate function. This will allow you to write a unit test for it and it will make the angelic_search function more readable.
i.e. prefix, suffix = find_prefix_and_suffix (....)